### DISCUSSION TITLE
accessing ios keychain and android shared preferences or accessing 'window' object

### DISCUSSION THREAD

**Discussion by mahendra331** [webdriverio/webdriverio#6916] [Q&A] [answered]
Hi, 
 We are looking to use webdriver io  for automating our Cordova Hybrid App. In our app, we use 2 factor authentication (via email). We save this authentication details in **keychain for iOS devices** and in **Shared preferences for Android devices**. Generally cordova injects plugins on 'window' object, we use one of the plugin to save these details calling '_window.plugin.save()_' As we can't use 2 factor authentication for tests, we are looking to save these details in Keychain / Shared preferences in 'before' hook so that app uses these directly without going through login process.
Do we have a way to save details in Keychain / Shared Preferences directly? Or do we have a way to access 'window' object in 'before' hook so that we can save the details using plugin? I tried this but we are getting 'window is not defined' error.
Can you please help me fixing this issue? 

Thanks for your help in advance.

**Comment by mahendra331** [comment, marked as answer]
I'll self answer this for others. We should be able to access window object using driver.execute method in webview context.

**Reply by wswebcreation** [reply]
@mahendra331 

This is possible with Appium and not really WDIO related. Just make sure you are in the WebView context. Be aware that an app can have multiple context and this can get difficult to select the correct one.

I gave a workshop about testing Hybrid apps that might make this _easier_.  Please check https://github.com/wswebcreation/automating-hybrid-apps and then the [answers](https://github.com/wswebcreation/automating-hybrid-apps/tree/main/tests/specs/answers) for some generic methods to do this.

Keep in mind that even Appium doesn't have access to the keychain/shared preferences. This is a secure part and no automation tool has access to writing to those stores.

Hope this helps
